### DISCUSSION TITLE
Error: No Destination found for a deployment for VM instance

### DISCUSSION THREAD

**Discussion by wagnerdracha** [apache/cloudstack#9580] [Q&A] [answered]
##### ISSUE TYPE

 * Other

##### COMPONENT NAME

Instances

##### CLOUDSTACK VERSION

CloudStack 4.19.1.0

##### OS / ENVIRONMENT

Ubuntu

##### SUMMARY

On the first time that I create a VM with default cloudstack template (CentOS), it worked perfectly!

But, I uploded a Ubuntu ISO e tried to create a VM with this ISO. After that, cloudstack show the error message: No Destination found for a deployment for VM instance.

I tried to delete the instances and create it again, but after that, I can't create another instance:
![image](https://github.com/user-attachments/assets/b56ad86d-87d9-4669-9933-c72ae87bd28a)

Looking at management-server.log, I see this message:
`2024-08-23 01:14:07,589 INFO  [o.a.c.a.c.u.v.DeployVMCmd] (API-Job-Executor-2:ctx-f71fceaf job-859 ctx-e1a8ecd5) (logid:b2d52dd2) No destination found for a deployment for VM instance {"id":108,"instanceName":"i-5-108-VM","type":"User","uuid":"d43d952b-a338-4225-955b-568e55b952ca"}: No destination found for a deployment for VM instance {"id":108,"instanceName":"i-5-108-VM","type":"User","uuid":"d43d952b-a338-4225-955b-568e55b952ca"}
2024-08-23 01:14:07,589 DEBUG [o.a.c.a.c.u.v.DeployVMCmd] (API-Job-Executor-2:ctx-f71fceaf job-859 ctx-e1a8ecd5) (logid:b2d52dd2) No destination found for a deployment for VM instance {"id":108,"instanceName":"i-5-108-VM","type":"User","uuid":"d43d952b-a338-4225-955b-568e55b952ca"}
com.cloud.exception.InsufficientServerCapacityException: No destination found for a deployment for VM instance {"id":108,"instanceName":"i-5-108-VM","type":"User","uuid":"d43d952b-a338-4225-955b-568e55b952ca"}Scope=interface com.cloud.dc.DataCenter; id=6`

I read that error could be a insufficient server capacity, but, you could see that I have resources:
![image](https://github.com/user-attachments/assets/2fd120e4-0a3a-4e48-b00e-4c59dbec6c4c)

To explain how the cloudstack is installed, I have a cloudstack installed on Ubuntu. Primary and Second storage shared on truenas NFS. The mysql was installed in another machine with Ubuntu and running a Galera cluster, and the KVM installed on Ubuntu.

Thnaks in advanced!

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by DaanHoogland** [comment]
@wagnerdracha , you'll need to look at the logs above the INFO that you shared to find the reason that no deployment plan could be created.

As the id of the VM is 108 there have been a lot created before. Check the following;
- is the ssvm running and the agent connected
- do the primary storages have capoacity
- are primary storages reachable for the hosts
- are there any VMs running in the hosts that are not known to ACS

**Reply by wagnerdracha** [reply]
I replied below.

**Reply by wagnerdracha** [reply]
Just to demonstrate, I post the instances deleted, but I can create another instance without delete one by one.
![image](https://github.com/user-attachments/assets/39eef7e8-82a5-4ffd-be76-2a3a1965cbfb)

**Comment by wagnerdracha** [comment]
@DaanHoogland, the reason to id of the VM is 108, it is because I had another problem before that cloudstack is not found the secondary storage. So, the ssvm tried to start and don't do it. But, I fixed this error.

- Now, the ssvm and the agent are working and connected. (is the ssvm running and the agent connected)
- The primary storage have capacity (looking at the image) (do the primary storages have capoacity)
![image](https://github.com/user-attachments/assets/4a996db1-e4ee-4944-9630-c0e31b33c5a2)
- I think that hosts was reach the primary storage, because I could started a VM yesterday (I would talk about that *). (are primary storages reachable for the hosts)
- Not had VMs running without ACS knowning. All the VM that I started was by ACS. (are there any VMs running in the hosts that are not known to ACS)

* Yesterday, I access the ACS with admin account and saw that existed many instances with the status Expunged. I selected all this instances and click on the trash button to delete all. Before that, the ACS say to me that all the selected instances was deleted, but all the instances was there with Expunged status. So, I tried to access all the instances individually and click on delete buttton, and it worked! I could delete the instances one by one.
After that, I created a new instance with the Ubuntu ISO and it started and worked!
So, I have a question. Why this worked now? If I had resources available (looking at the imagem above)

Thanks again!

**Reply by DaanHoogland** [reply, marked as answer]
Ok, this sounds like you had a lot of destroyed VMs that were not expunged yet. You can destroy and expunge in which case the space is freed immediately, or you can destroy and let garbage collection do the expunging in which case date will be kept around a configurable while in case you regret the destroying.

What probably happened is that you created and destroyed but did not expunge a lot of VMs and at a certain point you went beyond capacity. Afterthat you let the env rest a while and the garbage clean thread kicked in and expunged your VMs. You would not even have to delete the expunged VMs. The space would already have been freed.

**Reply by wagnerdracha** [reply]
Yes, I had a 10 VM destroyed and 1 VM running. But, my question is what resource capacity this VMs need to not get this error again? I ask this, because if you see the image above, you will see that my cpu, memory, primary and second storage had capacity. What I need to free or increase to not have this error? 
And thanks for your help!

**Reply by DaanHoogland** [reply]
@wagnerdracha I don't think you needed to free anything.
The primary storage was the problem but only because of the delay in freeing up the storage after destroying the VMs. In production you'll need to tune this to get optimal bang for you buck. In case you are expecting near full capacity use of your cloud with a strong turnover, you'll need to make sure destroyed VMs and disks get expunged sooner.

**Reply by wagnerdracha** [reply]
@DaanHoogland, reading your previous answer, the primary storage was the problem, but why? If the primary storage have free space. I don't understand yet.
Now, after I expurged the VMs, I could create a instance (how I said my reply before). But now, I try to create a new instance, but I got the same error again. But why if I have free resources?
Could you give a north to read/search/study to undertand what happening?

**Reply by DaanHoogland** [reply]
@wagnerdracha , I understand the confusion. It has to do with recoverability. please read https://docs.cloudstack.apache.org/en/latest/adminguide/virtual_machines.html#deleting-instance. It uses not much more words than I do, but maybe fromulates it a little better. Also type "expung" in the search box on the "Configuration" > "Global Settings" page to learn more about tuning this feature.

**Reply by wagnerdracha** [reply]
@DaanHoogland, thank you for your help! I think I understood what happened (I believe so). My primary storage has 50GB, so in the first example (1 CentOS and 1 Ubuntu instance) that worked, I had 10 GB of HD allocated for CentOS and 30 GB for Ubuntu. So CloudStack "allocated" 40 GB, leaving 10 GB free.

In the second example, I deleted the initial instances (CentOS and Ubuntu) and tried to create a new Ubuntu instance with 50 GB of HD, and it worked. However, when I tried to create another new Ubuntu instance with 50 GB of HD, it didn't work.

I thought this was possible because when I checked the Primary Storage dashboard, I saw that only 4.59% was used. However, CloudStack must allocate the GB for all instances created to ensure they will work correctly. I think that's how CloudStack works.

So, I created a new Primary Storage with 150 GB, and the problem was solved and I was able to create a new instance.

Thank you for your help!
